File: jetbrains-plugin/src/main/kotlin/com/asakii/plugin/services/LanguageAnalysisService.kt
package com.asakii.plugin.services

import com.intellij.openapi.components.Service
import com.intellij.openapi.components.service
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiElement
import com.intellij.psi.search.GlobalSearchScope

/**
 * 语言分析服务接口
 *
 * 提供跨语言的代码分析功能。使用 IntelliJ 的服务机制实现可选依赖：
 * - 默认实现 (NoopLanguageAnalysisService): 当 Java 插件不可用时使用，返回空结果
 * - Java 实现 (JavaLanguageAnalysisService): 当 Java 插件可用时自动覆盖，提供完整功能
 *
 * 通过 plugin.xml 和 plugin-withJava.xml 的 overrides="true" 机制自动切换实现。
 *
 * 使用方式：
 * ```kotlin
 * val service = LanguageAnalysisService.getInstance(project)
 * if (service.isAvailable()) {
 *     val inheritors = service.findClassInheritors(psiClass, scope)
 * }
 * ```
 */
interface LanguageAnalysisService {

    /**
     * 检查服务是否可用
     * @return true 如果语言分析功能可用（Java 插件已加载）
     */
    fun isAvailable(): Boolean

    /**
     * 检查元素是否是类（PsiClass）
     * @param element PSI 元素
     * @return true 如果是类
     */
    fun isClass(element: PsiElement): Boolean

    /**
     * 检查元素是否是方法（PsiMethod）
     * @param element PSI 元素
     * @return true 如果是方法
     */
    fun isMethod(element: PsiElement): Boolean

    /**
     * 搜索类的继承者
     * @param psiClass 类元素（必须是 PsiClass）
     * @param scope 搜索范围
     * @param deep 是否深度搜索（包含间接继承）
     * @return 继承者列表
     */
    fun findClassInheritors(
        psiClass: PsiElement,
        scope: GlobalSearchScope,
        deep: Boolean = true
    ): List<PsiElement>

    /**
     * 搜索覆盖的方法
     * @param psiMethod 方法元素（必须是 PsiMethod）
     * @param scope 搜索范围
     * @param deep 是否深度搜索
     * @return 覆盖方法列表
     */
    fun findOverridingMethods(
        psiMethod: PsiElement,
        scope: GlobalSearchScope,
        deep: Boolean = true
    ): List<PsiElement>

    /**
     * 获取 PsiShortNamesCache 实例
     * @return 缓存实例，不可用时返回 null
     */
    fun getShortNamesCache(): Any?

    /**
     * 获取所有类名
     * @param cache PsiShortNamesCache 实例
     * @return 类名数组
     */
    fun getAllClassNames(cache: Any?): Array<String>

    /**
     * 根据名称获取类
     * @param cache PsiShortNamesCache 实例
     * @param name 类名
     * @param scope 搜索范围
     * @return 匹配的类列表
     */
    fun getClassesByName(cache: Any?, name: String, scope: GlobalSearchScope): List<PsiElement>

    /**
     * 获取所有方法名
     * @param cache PsiShortNamesCache 实例
     * @return 方法名数组
     */
    fun getAllMethodNames(cache: Any?): Array<String>

    /**
     * 根据名称获取方法
     * @param cache PsiShortNamesCache 实例
     * @param name 方法名
     * @param scope 搜索范围
     * @return 匹配的方法列表
     */
    fun getMethodsByName(cache: Any?, name: String, scope: GlobalSearchScope): List<PsiElement>

    /**
     * 获取所有字段名
     * @param cache PsiShortNamesCache 实例
     * @return 字段名数组
     */
    fun getAllFieldNames(cache: Any?): Array<String>

    /**
     * 根据名称获取字段
     * @param cache PsiShortNamesCache 实例
     * @param name 字段名
     * @param scope 搜索范围
     * @return 匹配的字段列表
     */
    fun getFieldsByName(cache: Any?, name: String, scope: GlobalSearchScope): List<PsiElement>

    companion object {
        /**
         * 获取服务实例
         * IDE 会自动选择可用的实现（Java 或 Noop）
         */
        fun getInstance(project: Project): LanguageAnalysisService {
            return project.service<LanguageAnalysisService>()
        }
    }
}

File: vscode-extension/src/ide/services/languageAnalysisService.ts
import * as vscode from 'vscode';

/**
 * Symbol information representing a code element (class, method, field, etc.)
 */
export interface SymbolInfo {
    /** Symbol name */
    name: string;
    /** Symbol kind (class, method, field, etc.) */
    kind: vscode.SymbolKind;
    /** Location in the file */
    location: vscode.Location;
    /** Container name (e.g., class name for a method) */
    containerName?: string;
    /** Full qualified name */
    fullName?: string;
}

/**
 * Search scope for language analysis operations
 */
export interface SearchScope {
    /** Include workspace files */
    includeWorkspace: boolean;
    /** Include node_modules or dependencies */
    includeDependencies: boolean;
    /** Specific file patterns to include (glob patterns) */
    includePatterns?: string[];
    /** Specific file patterns to exclude (glob patterns) */
    excludePatterns?: string[];
}

/**
 * Default search scope - workspace only
 */
export const DEFAULT_SEARCH_SCOPE: SearchScope = {
    includeWorkspace: true,
    includeDependencies: false,
};

/**
 * Language Analysis Service Interface
 *
 * Provides cross-language code analysis functionality using VS Code LSP APIs.
 * This is the VS Code equivalent of JetBrains' LanguageAnalysisService.
 *
 * Implementation Strategy:
 * - Default implementation (NoopLanguageAnalysisService): Returns empty results when LSP is unavailable
 * - VSCode implementation (VSCodeLanguageAnalysisService): Uses VS Code Language Server Protocol APIs
 *
 * Usage:
 * ```typescript
 * const service = LanguageAnalysisService.getInstance();
 * if (service.isAvailable()) {
 *     const inheritors = await service.findClassInheritors(symbolUri, position);
 * }
 * ```
 */
export interface LanguageAnalysisService {
    /**
     * Check if the service is available
     * @returns true if language analysis features are available (LSP is active)
     */
    isAvailable(): boolean;

    /**
     * Check if a symbol at the given position is a class
     * @param uri Document URI
     * @param position Position in the document
     * @returns true if the symbol is a class/interface/type
     */
    isClass(uri: vscode.Uri, position: vscode.Position): Promise<boolean>;

    /**
     * Check if a symbol at the given position is a method/function
     * @param uri Document URI
     * @param position Position in the document
     * @returns true if the symbol is a method/function
     */
    isMethod(uri: vscode.Uri, position: vscode.Position): Promise<boolean>;

    /**
     * Find classes that inherit from the class at the given position
     * Uses vscode.executeImplementationProvider to find implementations
     *
     * @param uri Document URI containing the class
     * @param position Position of the class symbol
     * @param scope Search scope
     * @param deep Whether to include indirect inheritors (transitive)
     * @returns List of inheriting class locations
     */
    findClassInheritors(
        uri: vscode.Uri,
        position: vscode.Position,
        scope?: SearchScope,
        deep?: boolean
    ): Promise<vscode.Location[]>;

    /**
     * Find methods that override the method at the given position
     * Uses vscode.executeImplementationProvider to find implementations
     *
     * @param uri Document URI containing the method
     * @param position Position of the method symbol
     * @param scope Search scope
     * @param deep Whether to include indirect overrides (transitive)
     * @returns List of overriding method locations
     */
    findOverridingMethods(
        uri: vscode.Uri,
        position: vscode.Position,
        scope?: SearchScope,
        deep?: boolean
    ): Promise<vscode.Location[]>;

    /**
     * Get all class names in the workspace
     * Uses vscode.executeWorkspaceSymbolProvider
     *
     * @param scope Search scope
     * @returns Array of class names
     */
    getAllClassNames(scope?: SearchScope): Promise<string[]>;

    /**
     * Get classes by name
     * Uses vscode.executeWorkspaceSymbolProvider
     *
     * @param name Class name to search for
     * @param scope Search scope
     * @returns List of matching class symbols
     */
    getClassesByName(name: string, scope?: SearchScope): Promise<SymbolInfo[]>;

    /**
     * Get all method names in the workspace
     * Uses vscode.executeWorkspaceSymbolProvider
     *
     * @param scope Search scope
     * @returns Array of method names
     */
    getAllMethodNames(scope?: SearchScope): Promise<string[]>;

    /**
     * Get methods by name
     * Uses vscode.executeWorkspaceSymbolProvider
     *
     * @param name Method name to search for
     * @param scope Search scope
     * @returns List of matching method symbols
     */
    getMethodsByName(name: string, scope?: SearchScope): Promise<SymbolInfo[]>;

    /**
     * Get all field/property names in the workspace
     * Uses vscode.executeWorkspaceSymbolProvider
     *
     * @param scope Search scope
     * @returns Array of field names
     */
    getAllFieldNames(scope?: SearchScope): Promise<string[]>;

    /**
     * Get fields by name
     * Uses vscode.executeWorkspaceSymbolProvider
     *
     * @param name Field name to search for
     * @param scope Search scope
     * @returns List of matching field symbols
     */
    getFieldsByName(name: string, scope?: SearchScope): Promise<SymbolInfo[]>;

    /**
     * Get document symbols for a file
     * Uses vscode.executeDocumentSymbolProvider
     *
     * @param uri Document URI
     * @returns List of symbols in the document
     */
    getDocumentSymbols(uri: vscode.Uri): Promise<vscode.DocumentSymbol[]>;

    /**
     * Find all references to a symbol
     * Uses vscode.executeReferenceProvider
     *
     * @param uri Document URI
     * @param position Position of the symbol
     * @param includeDeclaration Whether to include the declaration itself
     * @returns List of reference locations
     */
    findReferences(
        uri: vscode.Uri,
        position: vscode.Position,
        includeDeclaration?: boolean
    ): Promise<vscode.Location[]>;

    /**
     * Go to definition of a symbol
     * Uses vscode.executeDefinitionProvider
     *
     * @param uri Document URI
     * @param position Position of the symbol
     * @returns Definition location(s)
     */
    findDefinition(
        uri: vscode.Uri,
        position: vscode.Position
    ): Promise<vscode.Location[]>;

    /**
     * Find type definition of a symbol
     * Uses vscode.executeTypeDefinitionProvider
     *
     * @param uri Document URI
     * @param position Position of the symbol
     * @returns Type definition location(s)
     */
    findTypeDefinition(
        uri: vscode.Uri,
        position: vscode.Position
    ): Promise<vscode.Location[]>;
}

/**
 * Singleton instance holder
 */
let instance: LanguageAnalysisService | null = null;

/**
 * Get the singleton instance of LanguageAnalysisService
 * Returns VSCodeLanguageAnalysisService if LSP is available, otherwise NoopLanguageAnalysisService
 */
export function getLanguageAnalysisService(): LanguageAnalysisService {
    if (!instance) {
        // Dynamic import to avoid circular dependencies
        // Will be set by the extension activation
        throw new Error('LanguageAnalysisService not initialized. Call setLanguageAnalysisService first.');
    }
    return instance;
}

/**
 * Set the singleton instance of LanguageAnalysisService
 * Called during extension activation
 */
export function setLanguageAnalysisService(service: LanguageAnalysisService): void {
    instance = service;
}

/**
 * Helper class with static getInstance method for compatibility with JetBrains API pattern
 */
export const LanguageAnalysisServiceFactory = {
    getInstance(): LanguageAnalysisService {
        return getLanguageAnalysisService();
    },

    setInstance(service: LanguageAnalysisService): void {
        setLanguageAnalysisService(service);
    }
};
